### PR TITLE
Add QoS to ROS2Topic.

### PR DESCRIPTION
**Pull request by calvertdw**
- ROS 2 common interfaces (default message types and Euclid type support) are now included with the library artifact. This is done so publishers and subscribers for the DDS common types including Euclid types can be created without external code/tricks.
- ROS2Publisher and QueuedROS2Publisher now have improved error output and exception handling
- Empty classes RealtimeROS2Publisher and RealtimeROS2Subscription have been removed. They weren't actually providing realtime functionality.
- ROS2Callback is now deprecated. You should use ROS2NodeInterface#createSubscription2 instead.
- APIs have been filled in to include ROS2Topic
- Creating nodes, publishers, and subscribers no longer throws IOException. This wasn't useful and just made the library frustrating to use.
- Quality of service (QoS) is now a setting of ROS2Topic. Each topic should define it's own QoS.
- RELIABLE is now an explicitly named ROS2QoSProfile option.
- DEFAULT quality of service (QoS) is now BEST_EFFORT instead of RELIABLE to help alleviate comms issues. This is only used if the topic does not specify the QoS to use. The default QoS can be set with the ROS_DEFAULT_QOS environment variable.
- The DEFAULT_NAMESPACE is now accessible via the public field in ROS2NodeBasics.

### CODE REVIEW

File: ihmc-ros2-library/build.gradle.kts [hunk 1]
```diff
@@ -15,6 +15,8 @@ mainDependencies {
    api("us.ihmc:ihmc-pub-sub:0.19.0")
    api("us.ihmc:ihmc-realtime:1.6.0")
    api("us.ihmc:ihmc-commons:0.32.0")
+
+   api("us.ihmc:ros2-common-interfaces:source")
```
Comment (ds58) [inline ihmc-ros2-library/build.gradle.kts:19]: How come this guy is needed?

File: ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2TopicNameTools.java [hunk 2]
```diff
@@ -231,27 +234,36 @@ public static <T> T newMessageInstance(Class<T> messageType)
    @SuppressWarnings({"unchecked", "rawtypes"})
    public static <T> TopicDataType<T> newMessageTopicDataTypeInstance(Class<T> messageType)
    {
+      if (messageType.equals(Pose3D.class))
```
Comment (ds58) [inline ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2TopicNameTools.java:270]: Why is this needed?

Comment (calvertdw) [reply ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2TopicNameTools.java:270]: If you want to send a Pose3D as a message itself. It's not possible without this. Most of the time we don't use that. Technically to be complete we'd add the cases for the other ones in `ros2-common-interfaces/src/main/java/us/ihmc/ros2/rosidl/geometry_msgs/msg/dds`
image

File: ihmc-ros2-library/build.gradle.kts (hunk 1 above)
Comment (calvertdw) [reply ihmc-ros2-library/build.gradle.kts:19]: That's for the change to `newMessageTopicDataTypeInstance`. To send the primitives as messages themselves, you need this. I think it was overlooked initially because most people don't do that.

Comment (calvertdw) [reply ihmc-ros2-library/build.gradle.kts:19]: I can't imagine the case where you'd somehow want to use the ihmc-ros2-library, but not the common msgs included with ROS and the Euclid type support. Seems silly to protect that use case.

File: ihmc-ros2-library/src/main/java/us/ihmc/ros2/QueuedROS2Publisher.java
Comment (calvertdw) [inline ihmc-ros2-library/src/main/java/us/ihmc/ros2/QueuedROS2Publisher.java:1]: Added the improvements we made in open robotics software and removed that version

File: ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2NodeBasics.java [hunk 3]
```diff
@@ -55,18 +65,27 @@ class ROS2NodeBasics implements ROS2NodeInterface
     * @param topicDataType       The topic data type of the message
     * @param publisherAttributes Publisher attributes created with @see{createPublisherAttributes}
     * @return a ROS 2 publisher
-    * @throws IOException if no publisher can be made
     */
    @Override
-   public <T> ROS2Publisher<T> createPublisher(TopicDataType<T> topicDataType, PublisherAttributes publisherAttributes) throws IOException
+   public <T> ROS2Publisher<T> createPublisher(TopicDataType<T> topicDataType, PublisherAttributes publisherAttributes)
```
Comment (calvertdw) [inline ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2NodeBasics.java:70]: I removed all the throws IOException that comes from pub sub and just throw the RuntimeExceptions here.

File: ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2Publisher.java
Comment (calvertdw) [inline ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2Publisher.java:1]: Added the improvements we made in open robotics software and removed that version

File: ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2QosProfile.java [hunk 4]
```diff
@@ -16,6 +18,11 @@
  */
 public class ROS2QosProfile
 {
+   public static ROS2QosProfile DEFAULT()
+   {
+      return BEST_EFFORT();
+   }
+
    public static ROS2QosProfile RELIABLE()
```
Comment (calvertdw) [inline ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2QosProfile.java:44]: RELIABLE is not explicitly named.

File: ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2QosProfile.java [hunk 5]
```diff
@@ -16,6 +18,11 @@
  */
 public class ROS2QosProfile
 {
+   public static ROS2QosProfile DEFAULT()
+   {
+      return BEST_EFFORT();
```
Comment (calvertdw) [inline ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2QosProfile.java]: DEFAULT is now BEST_EFFORT

File: ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2Topic.java [hunk 6]
```diff
@@ -58,6 +58,8 @@ public class ROS2Topic<T>
    private final String typeName;
    private final String suffix;
    private final Class<T> messageType;
+   /** Allows a QoS to be carried along with the topic to keep boilerplate to a minimum. */
+   private final ROS2QosProfile qos;
```
Comment (calvertdw) [inline ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2Topic.java:62]: qos is now a part of ROS2Topic

File: ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2TopicNameTools.java (hunk 2 above)
Comment (calvertdw) [reply ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2TopicNameTools.java:270]: I have filled this out for all the cases.

File: ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2QosProfile.java (hunk 5 above)
Comment (ds58) [reply ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2QosProfile.java]: Would it make sense to remove DEFAULT entirely? I think maybe it's better if you're explicit every time you use a QoS profile.

Comment (SylvainBertrand) [reply ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2QosProfile.java]: I personally like having a `DEFAULT` although I'm not against @ds58 suggestion.
Having said nothing, I think if we're to keep a `DEFAULT` it'd be nice to have the option to change it without having to clone this repo, so either looking for an environment variable or property.

**Review by SylvainBertrand — APPROVED** [review]
Overall it seems to me that this repo does maybe too much handholding which should preferably done in open robotics so it's easier to modify. However, I don't have strong opinion about it, so I'm good with this.

File: ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2QosProfile.java (hunk 5 above)
Comment (calvertdw) [reply ihmc-ros2-library/src/main/java/us/ihmc/ros2/ROS2QosProfile.java]: Oof it's a conundrum. I think in any case there'd probably be a default popping up in open robotics if we don't include it here, for those wanting simple APIs and they don't care (or don't yet know they care) about what happens under the hood.

I could be convinced to remove DEFAULT though if someone feels strongly.

In any case I've added the setting from a new ROS_DEFAULT_QOS environment variable. It'll print at the beginning one of the messages:
```
240403 12:30:56:645 [INFO] (ROS2QosProfile.java:34): Topics with unspecified QoS will use RELIABLE.
```
or
```
240403 12:29:51:646 [INFO] (ROS2QosProfile.java:39): Topics with unspecified QoS will use BEST_EFFORT.
```

**Review by SylvainBertrand — APPROVED** [review]

**Review by ds58 — APPROVED** [review]
